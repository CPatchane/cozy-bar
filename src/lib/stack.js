import internal from 'lib/stack-internal.js'
import client from "lib/stack-client"


/**
 * Reference to the current client depending
 * on which one has been initialized
 *
 * @private
 * @TODO We should set it to `undefined`
 * and throw an error when it is not initialized
 * Leaving with that default for today
 * so I do not need to rewrite deeply the tests
 * (they test some components without initializing
 * the stack client first, and rely on non-initialized
 * legacy client behaviour)
 */
let stack = internal

/**
 * Get the current stack client (legacy or cozy-client based)
 * based on which one has been initialized
 *
 * @returns {Object} functions to call the stack
 */
const current = function() {
  if (stack === undefined) {
    throw new Error("client not initialized in cozy-bar")
  }
  return stack
}

/**
 * Initializes the functions to call the cozy stack
 *
 * @function
 * @param {Object}  arg
 * @param {string}  arg.cozyURL - URL or domain of the stack
 * @param {boolean} arg.ssl     - Tells if we should use a secure protocol
 *                            required if cozyURL does not have a protocol
 * @param {string}  arg.token   - Access token for the stack
 * @param {Function} arg.onCreateApp 
 * @param {Function} arg.onDeleteApp
 * @returns {Promise}
 */
const init = function(options) {
  stack = internal
  return stack.init(options)
} 

const get = {
  app: (...args) => current().get.app(...args), 
  apps: (...args) => current().get.apps(...args), 
  context: (...args) => current().get.context(...args),
  storageData: (...args) => current().get.storageData(...args),
  iconProps: (...args) => current().get.iconProps(...args),
  cozyURL: (...args) => current().get.cozyURL(...args),
}

export default {
  init, 
  get, 
  updateAccessToken: (...args) => current().updateAccessToken(...args), 
  logout: (...args) => current().logout(...args), 
  cozyFetchJSON: (...args) => current().cozyFetchJSON(...args)
}
