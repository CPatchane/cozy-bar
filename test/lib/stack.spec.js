import internal from 'lib/stack-internal.js'
import client from 'lib/stack-client.js'
import stack from 'lib/stack.js'

const cozyURL = 'https://test.mycozy.cloud'
const token = 'mytoken'
const onCreateApp = function() {}
const onDeleteApp = function() {}

describe('stack proxy', () => {
  beforeAll(() => {
    jest.spyOn(internal, 'init').mockResolvedValue(undefined)
    jest.spyOn(internal.get, 'app').mockResolvedValue(undefined)

    jest.spyOn(client, 'init').mockResolvedValue(undefined)
    jest.spyOn(client.get, 'app').mockResolvedValue(undefined)
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  describe('when initialized with an cozyURL + token', () => {
    const params = { cozyURL, token, onCreateApp, onDeleteApp }

    beforeAll(() => {
      jest.clearAllMocks()

      stack.init(params)
      stack.get.app()
    })

    it('should call the internal stack init', () => {
      expect(internal.init).toHaveBeenCalled()
    })

    it('should not call the cozy-client stack init', () => {
      expect(client.init).not.toHaveBeenCalled()
    })

    it('should forward requests to the internal stack client', () => {
      expect(internal.get.app).toHaveBeenCalled()
    })

    it('should not forward requests to the cozy-client stack client', () => {
      expect(client.get.app).not.toHaveBeenCalled()
    })
  })

  describe('when initialized with a cozy-client instance', () => {
    const cozyClient = {
      getStackClient: () => {
        return {
          token: { token: 'mytoken' },
          uri: 'https://test.mycozy.cloud'
        }
      }
    }
    const params = {
      cozyClient,
      onCreateApp: function() {},
      onDeleteApp: function() {}
    }

    beforeAll(() => {
      jest.clearAllMocks()

      stack.init(params)
      stack.get.app()
    })

    it('should call the client stack init', () => {
      expect(client.init).toHaveBeenCalled()
    })

    it('should forward requests to the cozy-client stack client', () => {
      expect(client.get.app).toHaveBeenCalled()
    })
  })

  it('should throw error if trying to use the stack before init', () => {
    jest.clearAllMocks()
    const params = { cozyURL, token, onCreateApp, onDeleteApp }

    // TODO for now the stack is default to internal even before init
    // expect(() => {
    //   stack.getStack()
    // }).toThrowErrorMatchingSnapshot()
    stack.init(params)
    expect(() => {
      stack.getStack()
    }).not.toThrowError()
    expect(stack.getStack()).toMatchSnapshot()
  })
})
