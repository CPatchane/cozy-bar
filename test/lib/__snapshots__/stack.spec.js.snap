// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stack proxy should throw error if trying to use the stack before init 1`] = `"client not initialized in cozy-bar"`;
